### ISSUE TITLE
ACTNARS requires pre-registering all generic arguments.

### ISSUE THREAD

**Issue by cmeyertons** [autofac/Autofac#1027]
Autofac.Extras.Moq leverages ACTNARS to remove a lot of unnecessary boilerplate to new up test objects on demand.

for a closed generic, it is required to "pre-register" all of the generic type arguments, even if you have no need to inject one in a constructor anywhere.

https://github.com/autofac/Autofac.Extras.Moq/issues/22

**Comment by tillig** [comment]
What are you looking to have happen?

**Comment by tillig** [comment]
I came up with a repro for this:

```c#
using System;
using System.Collections.Generic;
using Autofac;
using Autofac.Features.ResolveAnything;

namespace autofacdemo
{
    class Program
    {
        public class Concrete<IDependency>
        {
            public Concrete()
            {
            }
        }

        public interface IDependency
        {
        }

        static void Main(string[] args)
        {
            var builder = new ContainerBuilder();
            builder.RegisterSource(new AnyConcreteTypeNotAlreadyRegisteredSource());
            using var container = builder.Build();
            // This will throw.
            container.Resolve<Concrete<IDependency>>();
        }
    }
}
```

Note the `Concrete` class has an interface-based generic type parameter but it also doesn't need any of that in the constructor. This should resolve fine, but it doesn't. Likely this is tied to the fix for #885 [over here](https://github.com/autofac/Autofac/blob/6071f84d13a1cb0df021d5464959309989d683b1/src/Autofac/Features/ResolveAnything/AnyConcreteTypeNotAlreadyRegisteredSource.cs#L98) and needs a little more robust handling.

```
Unhandled exception. Autofac.Core.Registration.ComponentNotRegisteredException: The requested service 'autofacdemo.Program+Concrete`1[[autofacdemo.Program+IDependency, autofacdemo, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]]' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency.
   at Autofac.ResolutionExtensions.ResolveService(IComponentContext context, Service service, IEnumerable`1 parameters)   at Autofac.ResolutionExtensions.Resolve(IComponentContext context, Type serviceType, IEnumerable`1 parameters)         at Autofac.ResolutionExtensions.Resolve[TService](IComponentContext context, IEnumerable`1 parameters)
   at Autofac.ResolutionExtensions.Resolve[TService](IComponentContext context)
   at autofacdemo.Program.Main(String[] args) in C:\dev\autofacdemo\Program.cs:line 26
```

I feel like we've hit this somewhere else, too, but I can't put my finger on where.

**Comment by tillig** [comment]
This looks like a duplicate of #925. There's already a failing/skipped test for this using abstract types. I'm going to close this and follow up there.
